src/main/kotlin/com/goodchoice/domain/review/persistence/ReviewRepository.kt

```diff
@@ -0,0 +1,165 @@
+}
+
+class ReviewJooqRepository(
+    private val db: DSLContext,
+    private val clock: Clock,
+    private val objectMapper: ObjectMapper
+) : ReviewRepository {
+    override fun create(
+        title: String,
+        author: Reference,
+        subject: Reference,
+        advantages: List<String>,
+        disadvantages: List<String>,
+        mark: Mark,
+        body: ReviewBodyCreationRequest
+    ): Reference {
+        val reviewId = UUID.randomUUID()
+        val reviewBodyId = UUID.randomUUID()
+
+        db.insertInto(REVIEW)
+            .set(REVIEW.ID, reviewId)
+            .set(REVIEW.TITLE, title)
+            .set(REVIEW.SUBJECT_ID, subject.id)
+            .set(REVIEW.MARK, mark.value)
+            .set(REVIEW.REVIEWER_ID, author.id)
+            .set(REVIEW.IS_SHOWN, true)
+            .execute()
+
+        db.insertInto(REVIEW_BODY)
+            .set(REVIEW_BODY.ID, reviewBodyId)
+            .set(REVIEW_BODY.CONTENT, body.content)
+            .set(REVIEW_BODY.CREATED_TIMESTAMP, clock.now())
+            .set(REVIEW_BODY.REVIEW_ID, reviewId)
+            .execute()
+
+        var ordering = 0
+        db.insertInto(
+            REVIEW_POINT,
+            REVIEW_POINT.CONTENT,
+            REVIEW_POINT.ORDERING,
+            REVIEW_POINT.REVIEW_ID,
+            REVIEW_POINT.TYPE
+        ).apply { advantages.forEach { values(it, ordering++, reviewId, ADVANTAGE) } }
+            .apply { disadvantages.forEach { values(it, ordering++, reviewId, DISADVANTAGE) } }
+            .execute()
+
+        return Reference(reviewId)
+    }
+
+    override fun vote(reviewId: UUID, issuerId: UUID, voteType: VoteType) {
+        db.insertInto(REVIEW_VOTE, REVIEW_VOTE.REVIEWER_ID, REVIEW_VOTE.REVIEW_ID, REVIEW_VOTE.TYPE)
+            .values(issuerId, reviewId, voteType.asJooqVoteType())
+            .onDuplicateKeyUpdate()
+            .set(REVIEW_VOTE.REVIEWER_ID, issuerId)
+            .set(REVIEW_VOTE.REVIEW_ID, reviewId)
+            .set(REVIEW_VOTE.TYPE, voteType.asJooqVoteType())
+            .execute()
+    }
+
+    override fun getVotesByReviewIdOrNull(reviewId: UUID, issuerId: UUID): ReviewVotes? {
+        return db.selectFrom(GET_REVIEW_VOTES_BY_ACTOR(issuerId))
+            .where(GET_REVIEW_VOTES_BY_ACTOR.REVIEW_ID.eq(reviewId)).fetchOne()?.map {
+                ReviewVotes(
+                    it[GET_REVIEW_VOTES_BY_ACTOR.UPVOTES_COUNT],
+                    it[GET_REVIEW_VOTES_BY_ACTOR.DOWNVOTES_COUNT],
+                    Vote(it[GET_REVIEW_VOTES_BY_ACTOR.OWN_VOTE].asVoteType())
+                )
+            }
+    }
+
+    override fun removeVote(reviewId: UUID, issuerId: UUID) {
+        db.deleteFrom(REVIEW_VOTE)
+            .where(REVIEW_VOTE.REVIEW_ID.eq(reviewId))
+            .and(REVIEW_VOTE.REVIEWER_ID.eq(issuerId))
+            .execute()
+    }
+
+    override fun getAllBySubject(
+        subjectId: UUID,
+        mark: Mark?,
+        issuerId: UUID?,
```

Comment:
Rename to `reviewerId`? According to our business rules, review can be created by reviewer only